weave-js/src/components/PagePanel.tsx

```diff
@@ -110,6 +112,42 @@ const JupyterControlsIcon = styled.div`
   }
 `;
 
+const HOST_SESSION_ID_COOKIE = `host_session_id`;
+
+function useEnablePageAnalytics() {
+  const history = useHistory();
+  const pathRef = useRef('');
+
+  useEffect(() => {
+    const options = {
+      context: {
+        hostSessionID: getCookie(HOST_SESSION_ID_COOKIE),
+      },
+    };
+
+    const unlisten = history.listen(location => {
+      const currentPath = `${location.pathname}${location.search}`;
+      const fullURL = `${window.location.protocol}//${window.location.host}${location.pathname}${location.search}${location.hash}`;
+      if (pathRef.current !== currentPath) {
+        trackPage({url: fullURL}, options);
+        pathRef.current = currentPath;
+      }
+    });
+
+    // Track initial page view
+    const initialPath = `${history.location.pathname}${history.location.search}`;
```

Comment:
This code block shares nearly identical code with the internals of the listen function. It might be DRYer to have define a single function that performs this action and call it in these two places